californium-core/src/main/java/org/eclipse/californium/core/network/stack/TCPBlockwiseLayer.java

```diff
@@ -0,0 +1,394 @@
+				int blockNum = block2.getNum();
+				int currentNum = blockNum;
+				boolean hasNextBlock = true;
+				Response block = null;
+
+				for (int i = 0; ((i < bertStepSize) && (hasNextBlock)); i++) {
+					block = status.getNextResponseBlock(block2);
+					hasNextBlock = block.getOptions().getBlock2().isM();
+					currentNum = currentNum + 1;
+					block2 = new BlockOption(BERT_SZX, hasNextBlock, currentNum);
+					responseBlockBuilder.put(block.getPayload());
+				}
+
+				if (status.isComplete()) {
+					// clean up blockwise status
+					LOGGER.log(Level.FINE, "peer has requested last block of blockwise transfer: {0}", status);
+					clearBlock2Status(key);
+				} else {
+					LOGGER.log(Level.FINE, "peer has requested intermediary block of blockwise transfer: {0}", status);
+				}
+
+				block.setPayload(getBody(responseBlockBuilder));
+				boolean m = block.getOptions().getBlock2().isM();
+				block.getOptions().setBlock2(BERT_SZX, m, blockNum);
+				exchange.setCurrentResponse(block);
+				lower().sendResponse(exchange, block);
+			}
+		} else {
+			super.handleInboundRequestForNextBlock(exchange, request, key, status);
+		}
+	}
+
+	/**
+	 * Send request to peers.Handling is done here only when BERT option is
+	 * enabled otherwise handling is done in the BlockwikseLayer.
+	 */
+	public void sendRequest(final Exchange exchange, final Request request) {
+
+		Request firstRequest = request;
+		Request requestToSend;
+		if (isBertEnabled) {
+
+			if (isTransparentBlockwiseHandlingEnabled()) {
+
+				BlockOption block2 = request.getOptions().getBlock2();
+				if (block2 != null && block2.getNum() > 0) {
+					isBertEnabled = false;
+					super.sendRequest(exchange, request);
+				} else if (requiresBlockwise(request)) {
+					// This must be a large POST or PUT request with BERT
+					// option.
+					ByteBuffer requestBuilder = ByteBuffer.allocate(bertStepSize * BERT_INT_BLOCK_SIZE);
+					boolean hasNextBlock = true;
+					KeyUri key = getKey(exchange, request);
+					firstRequest = startBlockwiseUpload(exchange, request);
+					requestBuilder.put(firstRequest.getPayload());
+
+					Block1BlockwiseStatus status = getBlock1Status(key);
+					for (int i = 1; ((i < bertStepSize) && (hasNextBlock)); i++) {
+						status.setCurrentNum(i);
+						requestToSend = status.getNextRequestBlock();
+						requestBuilder.put(requestToSend.getPayload());
+						hasNextBlock = requestToSend.getOptions().getBlock1().isM();
+					}
+					firstRequest.setPayload(getBody(requestBuilder));
+					firstRequest.getOptions().setBlock1(BERT_SZX, hasNextBlock, 0);
+				}
+			}
+			exchange.setCurrentRequest(firstRequest);
+			lower().sendRequest(exchange, firstRequest);
+		} else {
+			super.sendRequest(exchange, request);
+		}
+
+	}
+
+	/**
+	 * Send response to peers. Handling is done here only when BERT option is
+	 * enabled otherwise handling is done in the BlockwikseLayer.
+	 */
+	public void sendResponse(final Exchange exchange, final Response response) {
+
```

Comment:
`BlockwiseLayer.sendResponse` in 2.0.x contains some more checks and error handling, this is missing here.